src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -688,3 +688,53 @@ ReplicateAllReferenceTablesToNode(WorkerNode *workerNode)
 		}
 	}
 }
+
+
+/*
+ * NodeHasAllReferenceTablesReplicas returns whether the given worker node has reference
+ * table replicas. If there are no reference tables the function returns true.
+ *
+ * This function does not do any locking, so the situation could change immediately after,
+ * though we can only ever transition from false to true, so only "false" could be the
+ * incorrect answer.
+ *
+ * In the case where the function returns true because no reference tables exist
+ * on the node, a reference table could be created immediately after. However, the
+ * creation logic guarantees that this reference table will be created on all the
+ * nodes, so our answer was correct.
+ */
+bool
+NodeHasAllReferenceTableReplicas(WorkerNode *workerNode)
```

Comment:
Although I can't find a way to break it, the no-locking, the `AllReferenceTable` reference and the fact that we only check for the first reference table to be found on the worker node leaves a gap that I am not confident about.

What is the reason to not _temporarily_ get a lock on the colocation group that would prevent any shard group movement to happen while we check the reference table colocation group?

Reply:
I guess we'd be guarding against DROP TABLE between CitusTableTypeIdList and LoadShardIntervalList, which would result in an error, without loss of correctness. We probably have a similar race in various places. 

I'm not convinced taking a lock that conflicts with the DROP TABLE leads to better usability. We would be waiting arbitrarily long for transactions like `BEGIN; DROP TABLE IF EXISTS .. ; CREATE TABLE .. ; SELECT create_reference_table...; do stuff; COMMIT;` every time, while holding many other locks. 

We could perhaps write a loop if we want to solve this, but given how rarely this function gets called I would opt for simplicity.